sdk/storage/Azure.Storage.Common/src/Shared/LazyLoadingReadOnlyStream.cs

```diff
@@ -0,0 +1,240 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System;
+using System.Globalization;
+using System.IO;
+using System.Threading;
+using System.Threading.Tasks;
+using Azure.Core.Pipeline;
+
+namespace Azure.Storage.Shared
+{
+    internal class LazyLoadingReadOnlyStream<T> : Stream
+    {
+        /// <summary>
+        /// The current position within the blob or file.
+        /// </summary>
+        private long _position;
+
+        /// <summary>
+        /// Last known length of underlying blob or file.
+        /// </summary>
+        private long _length;
+
+        /// <summary>
+        /// The number of bytes in the last download call.
+        /// </summary>
+        private int _lastDownloadBytes;
+
+        /// <summary>
+        /// The number of bytes to download per call.
+        /// </summary>
+        private readonly int _bufferSize;
+
+        /// <summary>
+        /// The stream we will download to.
+        /// </summary>
+        private Stream _stream;
+
+        /// <summary>
+        /// Request conditions to send on the download requests.
+        /// </summary>
+        private readonly object _requestConditions;
+
+        /// <summary>
+        /// Async DownloadTo() function.
+        /// </summary>
+        private readonly Func<HttpRange, object, bool, CancellationToken, Task<Response<T>>> _downloadToAsyncFunc;
+
+        /// <summary>
+        /// Sync DownloadTo() function.
+        /// </summary>
+        private readonly Func<HttpRange, object, bool, CancellationToken, Response<T>> _downloadToFunc;
+
+        public LazyLoadingReadOnlyStream(
+            Func<HttpRange, object, bool, CancellationToken, Task<Response<T>>> downloadToAsyncFunc,
```

Comment:
Is it worth taking just one delegate and passing in the `bool async`?

Reply:
Fixed.